modules/core/arrow-annotations-processor/src/main/java/arrow/generic/ProductProcessor.kt

```diff
@@ -0,0 +1,84 @@
+package arrow.generic
+
+import com.google.auto.service.AutoService
+import arrow.common.utils.AbstractProcessor
+import arrow.common.utils.asClassOrPackageDataWrapper
+import arrow.common.utils.knownError
+import me.eugeniomarletti.kotlin.metadata.KotlinClassMetadata
+import me.eugeniomarletti.kotlin.metadata.extractFullName
+import me.eugeniomarletti.kotlin.metadata.isDataClass
+import me.eugeniomarletti.kotlin.metadata.kotlinMetadata
+import me.eugeniomarletti.kotlin.metadata.proto
+import org.jetbrains.kotlin.serialization.ProtoBuf
+import java.io.File
+import javax.annotation.processing.Processor
+import javax.annotation.processing.RoundEnvironment
+import javax.lang.model.SourceVersion
+import javax.lang.model.element.TypeElement
+
+import javax.lang.model.element.Element
+
+@AutoService(Processor::class)
+class ProductProcessor : AbstractProcessor() {
+
+    private val annotatedProduct = mutableListOf<AnnotatedGeneric>()
+
+    override fun getSupportedSourceVersion(): SourceVersion = SourceVersion.latestSupported()
+
+    override fun getSupportedAnnotationTypes() = setOf(productAnnotationClass.canonicalName)
+
+    override fun onProcess(annotations: Set<TypeElement>, roundEnv: RoundEnvironment) {
+        annotatedProduct += roundEnv
+                .getElementsAnnotatedWith(productAnnotationClass)
+                .map(this::evalAnnotatedProductElement)
+
+        if (roundEnv.processingOver()) {
+            val generatedDir = File(this.generatedDir!!, "").also { it.mkdirs() }
+            ProductFileGenerator(annotatedProduct, generatedDir).generate()
+        }
+    }
+
+    private fun productAnnotationError(element: Element, annotationName: String, targetName: String): String = """
+            |Cannot use $annotationName on ${element.enclosingElement}.${element.simpleName}.
+            |It can only be used on $targetName.""".trimMargin()
+
+    private fun evalAnnotatedProductElement(element: Element): AnnotatedGeneric = when {
+        (element.kotlinMetadata as? KotlinClassMetadata)?.data?.classProto?.isDataClass == true -> {
+            val elementClassData = getClassData(element)
+            val paramNames = getConstructorParamNames(element)
+            val typeNames = getConstructorTypesNames(element)
+            val properties = paramNames.zip(typeNames).map { Target(it.second,it.first) }
+            if (properties.size > 22)
+                knownError("${element.enclosingElement}.${element.simpleName} up to 22 constructor parameters is supported")
+            else
+                AnnotatedGeneric(element as TypeElement, elementClassData, properties)
+        }
+
+        else -> knownError(productAnnotationError(element, productAnnotationName, productAnnotationTarget))
+    }
+
+    private fun getConstructorTypesNames(element: Element): List<String> = element.kotlinMetadata
```

Comment:
shouldn't we extract all this utility functions to the commons package? getting the constructor type names seems quite important and it can probably be reused by many APs. Since we're moving on this direction where we use APs a lot it seems like a good idea to create our own infraestructure syntax to work with them. Same for the following methods.

Reply:
Done